sbi/inference/npse/npse.py

```diff
@@ -0,0 +1,612 @@
+# This file is part of sbi, a toolkit for simulation-based inference. sbi is licensed
+# under the Affero General Public License v3, see <https://www.gnu.org/licenses/>.
+import time
+from copy import deepcopy
+from typing import Any, Callable, Optional, Union
+from warnings import warn
+
+import torch
+from torch import Tensor, ones
+from torch.distributions import Distribution
+from torch.nn.utils.clip_grad import clip_grad_norm_
+from torch.optim.adam import Adam
+from torch.utils.tensorboard.writer import SummaryWriter
+
+from sbi import utils as utils
+from sbi.inference import NeuralInference, check_if_proposal_has_default_x
+from sbi.inference.posteriors import (
+    DirectPosterior,
+)
+from sbi.inference.posteriors.base_posterior import NeuralPosterior
+from sbi.inference.posteriors.score_posterior import ScorePosterior
+from sbi.neural_nets.estimators.score_estimator import ConditionalScoreEstimator
+from sbi.neural_nets.factory import posterior_score_nn
+from sbi.utils import (
+    RestrictedPrior,
+    check_estimator_arg,
+    handle_invalid_x,
+    npe_msg_on_invalid_x,
+    test_posterior_net_for_multi_d_x,
+    validate_theta_and_x,
+    warn_if_zscoring_changes_data,
+    x_shape_from_simulation,
+)
+from sbi.utils.sbiutils import ImproperEmpirical, mask_sims_from_prior
+
+
+class NPSE(NeuralInference):
+    def __init__(
+        self,
+        prior: Optional[Distribution] = None,
+        score_estimator: Union[str, Callable] = "mlp",
+        sde_type: str = "ve",
+        device: str = "cpu",
+        logging_level: Union[int, str] = "WARNING",
+        summary_writer: Optional[SummaryWriter] = None,
+        show_progress_bars: bool = True,
+        **kwargs,
+    ):
+        """Base class for Neural Posterior Score Estimation methods.
+
+        Instead of performing conditonal *density* estimation, NPSE methods perform
+        conditional *score* estimation i.e. they estimate the gradient of the log
+        density using denoising score matching loss. We not only estimate the score
+        of the posterior, but a family of distributions analogous to diffusion models.
+
+        NOTE: Single-round NPSE is currently the only supported mode.
+
+        Args:
+            prior: Prior distribution.
+            score_estimator: Neural network architecture for the score estimator. Can be
+                a string (e.g. 'mlp') or a callable that returns a neural network.
```

Comment:
can it be anything else then mlp?

Reply:
> could also be a resnet as in FMPE, no? We have this implemented and could re-use it from FMPE builders (-> future PR?).

The score_builder will currently raise a NotImplementedError for "resnet" (but I agree, we should add it for consistency). But in "neural_net" refactoring PR we should put all these nets in a joint file.